### PR TITLE
Revert "Revert "Support macros when cross-compiling (#7118)" (#7352)"

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Reverts apple/swift-package-manager#7352.

Modified to build tests for the host triple when any macros are added to test modules directly as dependencies.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Sources/Build/BuildDescription/ResolvedTarget+BuildDescription.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import struct Basics.AbsolutePath
+import struct PackageGraph.ResolvedTarget
+
+import SPMBuildCore
+
+extension ResolvedTarget {
+    func tempsPath(_ buildParameters: BuildParameters) -> AbsolutePath {
+        let suffix = buildParameters.suffix(triple: self.buildTriple)
+        return buildParameters.buildPath.appending(component: self.c99name + "\(suffix).build")
```
[the remark this answers is not in the document]
Comment (MaxDesiatov) [reply Sources/Build/BuildDescription/ResolvedTarget+BuildDescription.swift:21]:
```suggestion
        return buildParameters.buildPath.appending(component: "\(self.c99name)\(suffix).build")
```

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Sources/PackageGraph/Resolution/ResolvedTarget.swift [hunk 2]
```diff
@@ -144,24 +144,70 @@ public struct ResolvedTarget {
...
+                }
+            }
+            self.buildTriple = inferredBuildTriple
+        } else {
+            self.buildTriple = target.buildTriple
+        }
+        self.updateBuildTriplesOfDependencies()
+    }
+
+    mutating func updateBuildTriplesOfDependencies(forceUpdate: Bool = false) {
```
Comment (MaxDesiatov) [inline Sources/PackageGraph/Resolution/ResolvedTarget.swift:195]: remove `forceUpdate` as unused

File: Sources/PackageGraph/Resolution/ResolvedTarget.swift [hunk 3]
```diff
@@ -144,24 +144,70 @@ public struct ResolvedTarget {
...
+                switch dependency {
+                case .target(let targetDependency, _):
+                    if targetDependency.type == .macro {
+                        inferredBuildTriple = .tools
+                        break loop
+                    }
+                case .product(let productDependency, _):
+                    if productDependency.type == .macro {
+                        inferredBuildTriple = .tools
+                        break loop                    }
```
Comment (MaxDesiatov) [inline Sources/PackageGraph/Resolution/ResolvedTarget.swift:185]: clean up formatting

File: Sources/PackageGraph/Resolution/ResolvedPackage.swift [hunk 4]
```diff
@@ -11,6 +11,9 @@
 //===----------------------------------------------------------------------===//
 
 import Basics
+
+import struct OrderedCollections.OrderedDictionary
```
Comment (MaxDesiatov) [inline Sources/PackageGraph/Resolution/ResolvedPackage.swift:15]: Remove as unneeded

**Review by xedin — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
This needs performance testing before it can be merged.

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by rauhul** [comment]
EXCITING :D

**Comment by finagolfin** [comment]
@MaxDesiatov, thanks for working on this, looking forward to using it.

I tried using the official April 22 snapshot for ubi9 x86_64 to cross-compile the swift-syntax macro examples, after patching `swift-build` to apply your fix from #7493, but I'm seeing errors like this:
```
> cd swift-syntax/Examples/
> ~/swift-DEVELOPMENT-SNAPSHOT-2024-04-22-a-ubi9/usr/bin/swift build --build-tests --destination ~/swift-android-sdk/android-aarch64.json -Xlinker -rpath -Xlinker \$ORIGIN/swift-trunk-android-aarch64-2024-04-22-24-sdk/usr/lib/swift/android
Building for debugging...
error: Failed opening '/home/fina/swift-syntax/Examples/.build/aarch64-unknown-linux-android28/debug/index/store/v5/units/AddAsyncMacroTests.swift.o-1VZ0WNZLXJYVM': No such file or directory
```
Swift-foundation also doesn't build with similar errors, but other macro-free packages cross-compile fine. Are you seeing build errors like this when cross-compiling packages that have macros?

**Pull request merged**